docs/RFCS/00000000_cascade.md

```diff
@@ -0,0 +1,1210 @@
+
+To describe the algorithm, we're going to concentrate on a single step of the recursion, performing a lookup of exactly a single table with an incoming operation. Let's first assume we have the following setup:
+
+```text
+'NA 'DC  'DS  'UC 'US
+  \   \   |   /   /
+    \  \  |  /  /
+      \ \ | / /
+        \\|//
+          X
+        //|\\
+      / / | \ \
+    /  /  |  \  \
+  /   /   |   \   \
+ NA  DC   DS   UC  US
+```
+
+Each element in this is a table.  Table `X` is the table in which the operation will be occurring. The other tables are labeled according to what type of dependency it has to `X`, or `X` has to it. All dependencies flow from the top down, so table `X` references a column in table `'NA`. And table `NA` references a column in table `X`.  All the tables on the first row, the ones that start with a `'` have a column that `X` is referencing and all tables on the 3rd row that don't start with the `'` are referencing table `X`.
+
+All the tables besides `X` are named based on the type of reference they have to `X`. Note that in practice, there are two options for each relationship, `ON DELETE` and `ON UPDATE`, but for simplicity, we assume that the other operation is a `NO ACTION` or `RESTRICT`. The tables are labeled accordingly:
+
+- `NA`, `'NA` - `NO ACTION` or `RESTRICT`
+- `DC`, `'DC` - `DELETE CASCADE`
+- `DS`, `'DS` - `DELETE SET NULL` or `DELETE SET DEFAULT`
+- `UC`, `'UC` - `UPDATE CASCADE`
+- `US`, `'US` - `UPDATE SET NULL` or `UPDATE SET DEFAULT`
+
+So table `X` has a `DELETE CASCADE` reference to `'DC` and `DC` has a `DELETE CASCADE` reference to `X`. So if a row was deleted in `'DC`, `X` would possibly delete a row and then `DC` would possibly delete a row.
+
+There are 3 operations that can occur on a value in a table and we will be looking at each one individually.
+
+Cycle detection is important and required here as the referential integrity actions can form circular dependencies. How this is accomplished will be described in the sections describing where this algorithm is applied.
+
+###### Insert
+
+When inserting a row into table `X`, all the tables that it references must be fetched. This can be done without worrying what type of relationship it has and the walk will not cascade from those tables.
+
+So we require the fetching of tables `'NA`, `'DC`, `'DS`, `'UC` and `'US` and we don't required them to be walked any further.
+
+For tables that reference `X`, inserting will not affect them and their table descriptors do not need to be fetched. Inserts never cascade.
+
```

Comment:
BTW this traversal of the table descriptors is part of "planning". I say this because @knz is splitting planning fro, execution, and @jordanlewis is going to caches these descriptors with the plan for reuse of prepare statements.